File: reverse-nodes-in-k-group/reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    
    // create fake head 
    let fakeHead = new ListNode(0, head);
    // set fakehead.next to head
    let previous = fakeHead;
    // set previous to fakehead
    // set current to head
    let current = head;
    let lastHead = fakeHead;
    while (current !== null) {
    // loop until current is null
        let i = 1;
        let newHead = current;
        // set j to 0
        // set newHead to current
        while (i < k && newHead !== null) {
        // loop until j = k OR newHead is null
            newHead = newHead.next;
            // set newHead to newHead.next
            i++;
            // increment j
        }
        if (newHead === null) {
            break;
        }
        // if newhead is null
            // break loop
        while (newHead !== current) {
        // loop while newHead!=current
            let temp = current.next;
            // save current.next in temp var
            // set current.next to previous
            current.next = previous;
            // set previous to current
            previous = current;
            // set current to temp var
            current = temp;
        
        }
        current = current.next;
        newHead.next = previous;
        previous = lastHead.next;
        lastHead.next = newHead;
        previous.next = current;
        lastHead = previous;
    }
    return fakeHead.next
};

File: reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode fakeHead;
        fakeHead.next = head;
        ListNode* current = head;
        ListNode* previous = &fakeHead;
        while (current != NULL) {
            ListNode* tail = previous;
            ListNode* newHead = current;
            int i = 1;
            while (i < k && newHead != NULL) {
                newHead = newHead->next;
                i++;
            }
            if (newHead == NULL) {
                break;
            }
            while (newHead != current) {
                ListNode* temp = current->next;
                current->next = previous;
                previous = current;
                current = temp;
            }
            current = current->next;
            newHead->next = previous;
            previous = tail->next;
            tail->next = newHead;
            previous->next = current;
        }
        return fakeHead.next;
    }
};